Exit and liquidation feature

Aerilym opened the pull request.

yougotwill commented on apps/staking/components/StakedNode/NodeRequestExitButton.tsx:
```diff
@@ -0,0 +1,396 @@
+    writeError,
+    transactionError,
+    simulateEnabled,
+  } = useInitiateRemoveBLSPublicKey({
```
This hook has gotten quite big damn.

yougotwill commented on apps/staking/components/StakedNode/NodeRequestExitButton.tsx:
```diff
@@ -0,0 +1,396 @@
+  ERROR,
+}
+
+function QueryStatusInformation({
```
Yeah definitely should split these components up a bit

Aerilym replied:
Yeah for sure, see above

yougotwill reviewed (changes requested):
- I can't build the staking app because it's complaining about the `@ts-expect-error` in `tooltip.tsx`. I replaced it with a `@ts-ignore` and it builds
- sent-staking-js/test.ts is missing `contract_id`
- When I went to `/stake` on localhost:3000 I get the following error

```tsx
RangeError: 730.511935977951 can't be converted to BigInt because it isn't an integer
Source

../packages/sent-staking-js/test.ts (37:13) @ BigInt

  35 |   return {
  36 |     address: generateWalletAddress(),
> 37 |     amount: BigInt(Math.random() * 1000),
     |             ^
  38 |     reserved: BigInt(Math.random() * 1000),
  39 |     locked_contributions: [],
  40 |   };
```

Aerilym commented on apps/staking/app/mystakes/modules/ClaimTokensModule.tsx:
```diff
@@ -260,6 +271,16 @@ function ClaimTokensDialog({
 }) {
   const dictionary = useTranslations('modules.claim.dialog');
 
+  const claimRewardsArgs = useMemo(
```
I'm not entirely convinced its needed, but I had to make this change for the new contract call as part of exiting, it probably does nothing in this case but best case scenario its memoised and not triggering extra re-renders, worst case, it does nothing.

Aerilym commented on apps/staking/locales/en.json:
```diff
@@ -413,12 +415,55 @@
-      "unlockingTimerNotification": "Unlocking {time}",
-      "unlockingTimerDescription": "Time left until the node is unlocked. After unlocking, you can withdraw your stake.",
+      "requestingExitTimerNotification": "Exit available {time}",
+      "requestingExitDescription": "Time left until the node is available for exit. Return {relative_time} ({date}) to exit the node.",
```
yeah `{date}` is, the dynamic variable `{date}` is just being wrapped in parenthesis to make the string look like:
`Time left until the node is available for exit. Return in 14 days (Wednesday, 11 September 2024 at 11:23:52pm GMT+10) to exit the node.`

Aerilym commented on apps/staking/components/StakedNode/NodeRequestExitButton.tsx:
```diff
@@ -0,0 +1,396 @@
+  );
+}
+
+function RequestNodeExitContractWriteDialog({ node }: { node: StakedNode }) {
```
Yeah im looking to make a contract write dialog generic as this has quite similar code to 2 other places already

Aerilym commented on packages/contracts/hooks/ServiceNodeRewards.tsx:
```diff
@@ -210,3 +210,29 @@ export function useAddBLSPubKey({
+}): UseInitiateRemoveBLSPublicKeyReturn {
+  const chain = useChain();
+
+  const defaultArgs = useMemo(() => [BigInt(contractId ?? 0)] as [bigint], [contractId]);
```
It has to be `[bigint]` as that is the smart contract arg, as in, it is a literal single item array where that one item is a `bigint`. This isn't be the same as `Array<bigint>` which is a 0-∞ size array where each item is a `bigint`.

yougotwill replied:
I see. My mistake sorry. All good 👍

Aerilym commented on apps/staking/components/StakedNodeCard.tsx:
```diff
@@ -103,12 +119,16 @@ const isUnlocked = (node: StakedNode): node is UnlockedStakedNode =>
-  node.state === NODE_STATE.DECOMMISSIONED;
+  !!(
+    node.state === NODE_STATE.DECOMMISSIONED &&
+    'deregistrationDate' in node &&
+    node.deregistrationDate
```
`'deregistrationDate' in node` checks that `deregistrationDate` is a property on the `node` object, as in it checks if it is declared/exists. This then allows us to do `node.deregistrationDate` as the previous check was true so we know `deregistrationDate` is declared and we can now check for its thuthyness

yougotwill replied:
Cool

Aerilym commented on packages/ui/components/ui/tooltip.tsx:
```diff
@@ -9,6 +9,7 @@ import {
   useState,
 } from 'react';
+/** @ts-expect-error TS doesnt know what its talking about */
 import { useDebounce } from '@uidotdev/usehooks';
```
damn, ok fixed

Aerilym commented:
> * I can't build the staking app because it's complaining about the `@ts-expect-error` in `tooltip.tsx`. I replaced it with a `@ts-ignore` and it builds
> 
>     * sent-staking-js/test.ts is missing `contract_id`
> 
>     * When I went to `/stake` on localhost:3000 I get the following error
> 
> 
> ```tsx
> RangeError: 730.511935977951 can't be converted to BigInt because it isn't an integer
> Source
> 
> ../packages/sent-staking-js/test.ts (37:13) @ BigInt
> 
>   35 |   return {
>   36 |     address: generateWalletAddress(),
> > 37 |     amount: BigInt(Math.random() * 1000),
>      |             ^
>   38 |     reserved: BigInt(Math.random() * 1000),
>   39 |     locked_contributions: [],
>   40 |   };
> ```

Changes made. Ive also fixed those broken tests for you :)

yougotwill reviewed (approved):
LGTM 🚢
